simpleviewinc/keg-hub
# Shared task options and docker build cleanup

**Ticket**: [ZEN-463](https://jira.simpleviewtools.com/browse/ZEN-463)

[Semver-Status] minor

> These changes are dependent on [this PR](https://github.com/simpleviewinc/keg-hub/pull/92)
> Please review those changes first. Once they are merged, I will update the PR to point to the develop branch

## Context

* These changes are part of a bigger ticket to update how docker images are managed. 
* These changes were extracted to to allow for a smaller PR. Because of this; some functionality may not be used in this PR
* To allow better management of docker images, we need to improve how images are tagged

## Goal

* Allow tasks to share common options
  * Common options only need to be updated in one location
  * They can then be reused for tasks that use that option
* Improve docker build functionality
  * Add more build options
* Added better tagging functionality for images

## Updates

* `repos/keg-cli/src/constants/tagMap.js`
  * Constants used when parse task options into docker image tags
* `repos/keg-cli/src/tasks/*`
  * Updates some tasks to share common options between them
* `repos/keg-cli/src/utils/docker/buildDockerCmd.js`
  * Refactored the docker build command to work with image tag updates
* `repos/keg-cli/src/utils/docker/getDockerCmdArgs.js`
  * Renamed from `getDockerArgs` to better specify what the function does
* `repos/keg-cli/src/utils/docker/tags/*`
  * Added a number of build tag helpers to parse images tags dynamically
* `repos/keg-cli/src/utils/task/options/*`
  * Shared task options
  * This allows updating the options in one location, which then updates all tasks that use them

## Testing

**Setup**
* Pull this PR => `keg cli && keg pr 91`
* Install the latest deps => `yarn install`
* Clear out all your old docker images and cache
  * `keg d clean -f`
  * Ensure the `keg-base` image is removed
    * Run `keg di rm`
    * Then select the base image if it still exists

**Test 1**
* Run the command `keg base build`
  * This should build the base image without issue
* Run the command `keg components build`
  * This should build the `keg-components` image without issue
* Run the command `keg evf build`
  * This should build the `tap` image without issue

**Test 2**
* Remove all the image images built in **Test 1**
* Run the command `keg components start`
  * This should build the `keg-base` && `keg-components` images without issue
  * It should then start the `keg-components` container
  * Ensure keg-components runs as expected
* Run the command `keg evf start`
  * This should build only the `tap` image without issue, because the base image was built in the last step
  * It should then start the `events force tap` container
  * Ensure the tap events force runs as expected

**Test 3**
* Run `keg evf pack`
  * This tests to ensure the `docker package` workflow still works properly
  * Ensure the package is created and pushed as expected


**Test 4**
* Run `keg evf pack run package=docker.pkg.github.com/simpleviewinc/keg-packages/tap:zen-468-pressable-update`
  * This tests to ensure the `docker provider pull` task still works as expected
  * Ensure the package runs as expected

**Test 5**
* Run the command `keg cli test`
  * Ensure all cli tests pass
